pil-analyzer/src/pil_analyzer.rs

```diff
@@ -257,6 +257,113 @@ impl PILAnalyzer {
+            }
+        });
+
+        for (sr, name, fields) in structs_exprs {
+            if let Some((
+                _,
+                Some(FunctionValueDefinition::TypeDeclaration(TypeDeclaration::Struct(
+                    struct_decl,
+                ))),
+            )) = self.definitions.get(&name)
+            {
+                let declared_fields: HashSet<_> =
+                    struct_decl.fields.iter().map(|f| &f.name).collect();
+                let used_fields: HashSet<_> = fields.iter().map(|f| &f.name).collect();
+
+                errors.extend(
+                    fields
+                        .iter()
+                        .filter(|f| !declared_fields.contains(&f.name))
+                        .map(|f| {
+                            sr.with_error(format!(
+                                "Struct '{name}' has no field named '{}'",
+                                f.name
+                            ))
+                        }),
+                );
+
+                errors.extend(declared_fields.difference(&used_fields).map(|&f| {
+                    sr.with_error(format!("Missing field '{f}' in initializer of '{name}'",))
+                }));
+
+                let duplicate_fields: Vec<_> =
+                    fields.iter().map(|f| &f.name).duplicates().collect();
+
+                errors.extend(
+                    duplicate_fields
+                        .into_iter()
+                        .map(|f| sr.with_error(format!("Field '{f}' specified more than once"))),
+                );
+            } else {
+                errors.push(sr.with_error(format!("Struct '{name}' has not been declared")));
+            }
+
+            visited_structs.insert(name);
+        }
+
+        let structs_decls = self.definitions.iter().filter(|(_, (_, def))| {
+            matches!(
+                def,
+                Some(FunctionValueDefinition::TypeDeclaration(
+                    TypeDeclaration::Struct(_)
+                ))
+            )
+        });
+
+        for (name, (symbol, def)) in structs_decls {
+            if !visited_structs.contains(name) {
+                errors.push(
+                    symbol
+                        .source
+                        .with_error(format!("warning: struct '{name}' is never constructed")), // TODO: Warningx
+                );
+            } else if let Some(FunctionValueDefinition::TypeDeclaration(TypeDeclaration::Struct(
+                struct_decl,
+            ))) = def
+            {
+                let duplicate_declaration_fields: Vec<_> = struct_decl
+                    .fields
+                    .iter()
+                    .map(|f| &f.name)
+                    .duplicates()
+                    .collect();
+
+                errors.extend(duplicate_declaration_fields.into_iter().map(|f| {
+                    symbol
+                        .source
+                        .with_error(format!("Field '{f}' is already declared",))
```

Comment:
Declared twice? Or do we have the source reference directly on the second field declaration?

Reply:
No, the source is for the complete struct. Maybe `is declared more than once` then?